src/Reserve.sol

```diff
@@ -1,258 +0,0 @@
-        plugins[token] = newPlugin;
-        uint256 amt = deposited[token];
-        if (oldPlugin != NO_PLUGIN) {
-            oldPlugin.beforeEnd(token, amt);
-        }
-        _transfer(token, _pluginAddr(oldPlugin), _pluginAddr(newPlugin), amt);
-        if (newPlugin != NO_PLUGIN) {
-            newPlugin.afterStart(token, amt);
-        }
-        emit PluginSet(msg.sender, token, oldPlugin, newPlugin, amt);
-    }
-
-    /// @notice Deposits funds into the reserve.
-    /// The reserve will `transferFrom` `amt` tokens from the `from` address.
-    /// Callable only by a current user.
-    /// @param token The used token.
-    /// @param from The address from which funds are deposited.
-    /// @param amt The deposited amount.
-    function deposit(IERC20 token, address from, uint256 amt) public override onlyUser {
-        IReservePlugin plugin = plugins[token];
-        require(from != address(plugin) && from != address(this), "Reserve: deposition from self");
-        deposited[token] += amt;
-        _transfer(token, from, _pluginAddr(plugin), amt);
-        if (plugin != NO_PLUGIN) {
-            plugin.afterDeposition(token, amt);
-        }
-        emit Deposited(msg.sender, token, from, amt);
-    }
-
-    /// @notice Withdraws funds from the reserve.
-    /// The reserve will transfer `amt` tokens to the `to` address.
-    /// Only funds previously deposited can be withdrawn.
-    /// Callable only by a current user.
-    /// @param token The used token.
-    /// @param to The address to which funds are withdrawn.
-    /// @param amt The withdrawn amount.
-    function withdraw(IERC20 token, address to, uint256 amt) public override onlyUser {
-        uint256 balance = deposited[token];
-        require(balance >= amt, "Reserve: withdrawal over balance");
-        deposited[token] = balance - amt;
-        IReservePlugin plugin = plugins[token];
-        if (plugin != NO_PLUGIN) {
-            plugin.beforeWithdrawal(token, amt);
-        }
-        _transfer(token, _pluginAddr(plugin), to, amt);
-        emit Withdrawn(msg.sender, token, to, amt);
-    }
-
-    /// @notice Withdraws funds from the reserve or a plugin.
-    /// The reserve will transfer `amt` tokens to the `to` address.
-    /// The function doesn't update the deposited amount counter.
-    /// If used recklessly, it may cause a mismatch between the counter and the actual balance
-    /// making valid future calls to `withdraw` or `setPlugin` fail due to lack of funds.
-    /// Callable only by the current owner.
-    /// @param token The used token.
-    /// @param plugin The plugin to withdraw from.
-    /// It doesn't need to be registered as a plugin for `token`.
-    /// Pass `NO_PLUGIN` to withdraw directly from the reserve balance.
-    /// @param to The address to which funds are withdrawn.
-    /// @param amt The withdrawn amount.
-    function forceWithdraw(IERC20 token, IReservePlugin plugin, address to, uint256 amt)
```

Comment:
So, we decided to completely remove the `forceWithdraw` functionality?

Reply:
Yes, it was needed first and foremost for collecting income from plugins by shaving off tokens without touching the internal counter. For extreme cases like emergency withdrawal after a total collapse of the protocol we'll need to update the contract and change its logic to do some kind of a refunding or at least draining to a given address. Do you think that we should have the draining function available at all times for the owner to call?